### PR TITLE
feat: nacos discovery support namespace and group and support multiple host

### PR DESCRIPTION
**Pull request by Demogorgon314**
### What this PR does / why we need it:
#4306
support multiple host
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/discovery/nacos.lua [hunk 1]
```diff
@@ -157,9 +157,23 @@ local function get_token_param(base_uri, username, password)
     return '&accessToken=' .. data.accessToken
 end
 
+local function get_group_and_namespace_param(service_info)
```
Comment (spacewander) [inline apisix/discovery/nacos.lua:160]: Need to encode the args.

File: docs/en/latest/discovery/nacos.md [hunk 2]
```diff
@@ -99,3 +99,97 @@ The format response as below:
   "action": "set"
 }
 ```
+
+example of routing a request with a URL of "/nacosWithNamespaceIdAndGroupId/*" to a service which name, namespaceId, groupName "http://192.168.33.1:8848/nacos/v1/ns/instance/list?serviceName=APISIX-NACOS&groupName=test&namespaceId=test" and use nacos discovery client in the registry :
```
Comment (spacewander) [inline docs/en/latest/discovery/nacos.md:103]: This first letter should be uppercase.

File: docs/en/latest/discovery/nacos.md [hunk 3]
```diff
@@ -99,3 +99,97 @@ The format response as below:
...
+      },
+      "priority": 0,
+      "uri": "\/nacosWithNamespaceIdAndGroupId\/*"
+    }
+  },
+  "action": "set"
+}
+```
+
+example to use other nacos host
```
Comment (spacewander) [inline docs/en/latest/discovery/nacos.md:148]: In what situation we need to switch Nacos address in the per upstream level?

File: apisix/schema_def.lua [hunk 4]
```diff
@@ -389,6 +389,24 @@ local upstream_schema = {
             description = "discovery type",
             type = "string",
         },
+        discovery_host = {
```
Comment (spacewander) [inline apisix/schema_def.lua:392]: Better to use `nacos_address` instead.
All these 3 fields should go under a separate field called `discovery_args`, as it is nacos specific.
And we need to check them only when discovery_type `nacos` is used.

File: apisix/discovery/nacos.lua [hunk 5]
```diff
@@ -208,7 +222,12 @@ local function iter_and_add_service(services, values)
         end
 
         if up.discovery_type == 'nacos' then
-            core.table.insert(services, up.service_name)
+            core.table.insert(services, {
+                service_name=up.service_name,
```
Comment (spacewander) [inline apisix/discovery/nacos.lua:226]: Need space around the `=`

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
Test must be added for the new feature.

**Comment by spacewander** [comment]
My suggestion is that we need to split the PR into steps:
1. add namespace
2. add group
3. add multiple host

So that the PR can be merged quickly.

**Comment by Demogorgon314** [comment]
> My suggestion is that we need to split the PR into steps:
> 
> 1. add namespace
> 2. add group
> 3. add multiple host
> 
> So that the PR can be merged quickly.

Thx I will close this pr and open another

**Pull request closed without merging**